File: AEDs/AEDs_II/Trabalhos/TrabalhosPraticos/TP04/TP04Q09/TP04Q09.c
#include "../Libs/LibsC/LibTP04.h"
#include "../Libs/LibsC/ListaDupla.h"

// clear && gcc TP04Q09.c && ./a.out < pub.in > result.txt

typedef struct HashInd {

	int tabelaLen;
	ListaDupla* array;

	void (*Inserir) (Jogador, struct HashInd);
	bool (*Pesquisar) (Jogador, Log* log, struct HashInd);
	void (*Mostrar) (struct HashInd);
	void (*Close) (struct HashInd);

} HashInd;

HashInd newHashInd();
int Hash(Jogador, HashInd);
void InserirHashInd(Jogador, HashInd);
bool PesquisarHashInd(Jogador, Log*, HashInd);
void MostrarHashInd(HashInd);
void CloseHashInd(HashInd);

int main() {

	BD BD = newBD(BD_SIZE);
	BD.ImportDataBase("../tmp/players.csv", &BD);

	HashInd hash = newHashInd(25);

	char inputPUBIN[STR_MAX_LEN];

	while (strcmp(readStr(0, inputPUBIN), "FIM")) {
		int id = atoi(inputPUBIN);
		hash.Inserir(BD.Get(id, BD), hash);
	}

	// hash.Mostrar(hash);

	Timer timer = newTimer();
	Log log = newLog();

	BD.OrdenarPorNome(&BD); // Para possibilitar uma pesquisa binária.

	timer.Start(&timer);
	while (strcmp(readStr(0, inputPUBIN), "FIM")) {
		Jogador pesquisado = BD.PesquisarPorNome(inputPUBIN, &log, BD);
		bool resultado = hash.Pesquisar(pesquisado, &log, hash);
		printf("%s %s\n", inputPUBIN, resultado ? "SIM" : "NAO");
	}
	timer.Stop(&timer);

	log.RegistroPesquisa("794989_hashIndireta.txt", timer, log);

	hash.Close(hash);
	BD.Close(&BD);
}

HashInd newHashInd(int tabelaLen) {

	HashInd hash;

	hash.tabelaLen = tabelaLen;
	hash.array = malloc(tabelaLen * sizeof(ListaDupla));

	for (int i = 0; i < tabelaLen; i++) {
		hash.array[i] = newListaDupla();
	}

	hash.Inserir = InserirHashInd;
	hash.Pesquisar = PesquisarHashInd;
	hash.Mostrar = MostrarHashInd;

	hash.Close = CloseHashInd;

	return hash;
}

int Hash(Jogador jogador, HashInd hash) {
	return jogador.altura % hash.tabelaLen;
}

void InserirHashInd(Jogador jogador, HashInd hash) {
	int pos = Hash(jogador, hash);
	hash.array[pos].InserirFim(jogador, &hash.array[pos]);
}

bool PesquisarHashInd(Jogador jogador, Log* log, HashInd hash) {
	int pos = Hash(jogador, hash);
	return hash.array[pos].Pesquisar(jogador, log, hash.array[pos]);
}

void MostrarHashInd(HashInd hash) {
	puts("----- Tabela Hash -----");
	for (int i = 0; i < hash.tabelaLen; i++) {
		hash.array[i].Mostrar(hash.array[i]);
	}
}

void CloseHashInd(HashInd hash) {

	for (int i = 0; i < hash.tabelaLen; i++) {
		hash.array[i].Close(&hash.array[i]);
	}

	free(hash.array);
}

File: AEDs/AEDs_II/Trabalhos/TrabalhosPraticos/TP04/TP04Q09/TP04Q09.java
package TP04.TP04Q09;

import TP04.Libs.LibsJava.Lib;
import TP04.Libs.LibsJava.Lib.BD;
import TP04.Libs.LibsJava.Lib.Log;
import TP04.Libs.LibsJava.Lib.Timer;
import TP04.Libs.LibsJava.Lib.Jogador;
import TP04.Libs.LibsJava.Lista;

// clear && javac -cp ../.. TP04Q09.java && java -cp ../.. TP04Q09.java < pub.in > result.txt

class TP04Q09 {

	static class HashInd {

		private Lista[] array;
		public final int tabelaLen;

		public HashInd(int tabelaLen) {
			this.tabelaLen = tabelaLen;
			this.array = new Lista[tabelaLen];
			for (int i = 0; i < tabelaLen; i++) {
				array[i] = new Lista();
			}
		}

		public void Mostrar() throws Exception {
			System.out.println("----- Tabela Hash -----");
			for (int i = 0; i < tabelaLen; i++) {
				array[i].Mostrar();
			}
		}

		private int hash(Jogador jogador) {
			return jogador.getAltura() % tabelaLen;
		}

		public void Inserir(Jogador jogador) {
			int pos = hash(jogador);
			array[pos].InserirFim(jogador);
		}

		boolean CompareToStr(String str1, String str2, Log log) {
			log.incrementarComparacoes();
			return str1.equals(str2);
		}

		public boolean Pesquisar(Jogador jogador, Log log) {
			int pos = hash(jogador);
			return array[pos].Pesquisar(jogador, log);
		}

	}

	public static void main(String[] args) throws Exception {

		BD BD = new BD("../tmp/players.csv");
		HashInd hash = new HashInd(25);

		String inputPUBIN = new String();
		
		while (!(inputPUBIN = Lib.readStr()).equals("FIM")) {
			int id = Integer.parseInt(inputPUBIN);
			hash.Inserir(BD.Get(id));
		}

		Log log = new Log();
		Timer timer = new Timer();

		BD.OrdenarPorNome(); // Para possibilitar uma pesquisa binária.

		timer.Start();
		while (!(inputPUBIN = Lib.readStr()).equals("FIM")) {
			Jogador pesquisado = BD.PesquisarPorNome(inputPUBIN, log);
			boolean resultado = hash.Pesquisar(pesquisado, log);
			System.out.printf("%s %s\n", inputPUBIN, resultado ? "SIM" : "NAO");
		}
		timer.Stop();

		log.RegistroPesquisa("794989_hashIndireta.txt", timer);
	}
}